### PR TITLE
New types for node-jose

### PR DESCRIPTION
**Pull request by Drynwynn**
Please fill in this template.

- [x] Make your PR against the `master` branch.
- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `tsc` without errors.
- [x] Run `npm run lint package-name` if a `tslint.json` is present.

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not provide its own types, and you can not add them.
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `npm run new-package package-name`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, and `strictNullChecks` set to `true`.

### CODE REVIEW

File: node-jose/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,1380 @@
...
+        data: Buffer, // the data passed into the verify function
+        mac: Buffer, // the signature for `data`
+        valid: boolean // whether the signature matches the data
+    }
+
+    /**
+     * Run the decryption algorithm given by `alg`, if it is supported in the current
+     * implementation.
+     *
+     * @param {string} alg The algorithm to use
```
Comment (andy-ms) [inline node-jose/index.d.ts]: Don't need type declarations in jsdoc, that's what TypeScript is for!

File: node-jose/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,1380 @@
...
+        protected: string[],
+        key: JWK.Key,
+        payload: Buffer,
+        plaintext: Buffer
+    };
+
+    /**
+     * The options for {@link createEncrypt}.
+     *
+     * @property {boolean|string} zip Determines the compression algorithm to
```
Comment (andy-ms) [inline node-jose/index.d.ts]: Move documentation to the properties theirselves, i.e.:
```ts
/** Describe zip */
zip?: boolean | string;
```

File: node-jose/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,1380 @@
...
+        /**
+         * Indicates the compression algorithm applied to the plaintext
+         * before it is encrypted.  The possible values are:
+         *
+         * + **`"DEF"`**: Compress the plaintext using the DEFLATE algorithm.
+         * + **`""`**: Do not compress the plaintext.
+         *
+         * @type {string}
+         * @memberOf Encrypter
+         * @readonly
```
Comment (andy-ms) [inline node-jose/index.d.ts]: TypeScript provides a `readonly` modifier so you don't need jsdoc for this.

**Review by ghost — COMMENTED** [review]
Needs a `tslint.json` containing `{ "extends": "../tslint.json" }`

File: node-jose/index.d.ts (hunk 3 above)
Comment (sandersn) [inline node-jose/index.d.ts]: Also the type should be `"DEF" | ""`, which is more specific than just `string`.

File: node-jose/index.d.ts (hunk 2 above)
Comment (sandersn) [inline node-jose/index.d.ts]: The `string`s for this type should be broken into their allowed values, eg `zip: boolean | "DEF" | ""`.

File: node-jose/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,1380 @@
...
+     * @property {string} format Determines the serialization format of the
+     *           output.  Expected to be `"general"` for general JSON
+     *           Serialization, `"flattened"` for flattened JSON Serialization,
+     *           or `"compact"` for Compact Serialization (default is
+     *           **`"general"`**).
+     * @property {boolean} compact Determines if the output is the Compact
+     *           serialization (`true`) or the JSON serialization (**`false`**,
+     *           the default).
+     * @property {string} contentAlg The algorithm used to encrypt the plaintext
+     *           (default is **`"A128CBC-HS256"`**).
```
Comment (sandersn) [inline node-jose/index.d.ts]: it's not clear whether `contentAlg` is open-ended, but even if it is `contentAlg: "A128CBC-HS256" | string` is better documentation than just string, even though they are technically equivalent.

File: node-jose/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,1380 @@
...
+/**
+ * Tools for creating and manipulating keys and keystores.
+ */
+export module JWK {
+
+    /**
+     * The following constants are used to select the different modes for
+     * operation, their names are self-explanatory.
+     */
+    const MODE_DECRYPT: string;
```
Comment (sandersn) [inline node-jose/index.d.ts]: these should probably be `const MODE_DECRYPT: "MODE_DECRYPT";` and then have a union type tying them together for later use: `type Mode = MODE_DECRYPT | MODE_ENCRYPT | ...`

File: node-jose/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,1380 @@
...
+         * @param {String} name The name of the parameter
+         * @param {Boolean} [isPrivate=false] `true` if private parameters should
+         *        be checked.
+         * @returns {any} The value of the named parameter, or undefined if
+         *          it is not present.
+         *
+         *
+         * @memberOf Key
+         */
+        get: (name: string, isPrivate?: boolean) => any;
```
Comment (sandersn) [inline node-jose/index.d.ts]: from the jsdoc, I think the second parameter should be `isPrivate = false`

**Review by sandersn — CHANGES_REQUESTED** [review]

File: node-jose/index.d.ts (hunk 6 above)
Comment (andy-ms) [inline node-jose/index.d.ts]: You can't have an initializer in a declaration file.

**Comment by andy-ms** [comment]
We've recently updated the repository structure.
All packages are now in a `types` subdirectory.
Please rebase this pull request onto the latest `master` branch.

**Comment by RyanCavanaugh** [comment]
@sandersn can you update your review status please?

**Comment by sandersn** [comment]
My comments aren't addressed, but it would be better to merge at this point so at least the definitions are available.

**Comment by blakeembrey** [comment]
This PR wasn't fixed up according to @andy-ms' comments, so you have one random definition in the root of the project now (https://github.com/DefinitelyTyped/DefinitelyTyped/tree/master/node-jose).

**Comment by Drynwynn** [comment]
Sorry for the delay all.

I'll review the comments (thanks for the feedback) and fix up the structure in another PR to fix what was added in this one.

**Comment by sandersn** [comment]
@sahlouls It was removed because the types needed more work: See https://github.com/DefinitelyTyped/DefinitelyTyped/pull/15953

**Pull request merged**